pkg/bigquery/sanitize.go

```diff
@@ -0,0 +1,98 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bigquery implements helpers for using BigQuery.
+package bigquery
+
+import (
+	"fmt"
+	"regexp"
+	"unicode/utf8"
+)
+
+const tableNameMaxUTF8Bytes = 1024
+
+// Start with lowercase, middle is lowercase, number or hyphen, cannot end in
+// hyphen. 6-30 characters in length (start, 4-28 middle, end).
+var projectIDMatcher = regexp.MustCompile(`^[a-z][a-z0-9\-]{4,28}[a-z0-9]$`)
+
+// Lowercase and uppercase letters and underscores. Max 1024 characters.
+// regexp only allows 1000 repetitions, so had to manually repeat.
+const datasetIDRegex = `^[a-zA-Z0-9_]{1,512}[a-zA-Z0-9_]{0,512}$`
+
+var datasetIDMatcher = regexp.MustCompile(datasetIDRegex)
+
+// Unicode characters in category L (letter), M (mark), N (number),
+// Pc (connector, including underscore), Pd (dash), Zs (space). Max 1024
+// characters. 1024 is actually UTF-8 bytes from experimentation.
+// UTF-8 length check will be done separately.
+// regexp only allows 1000 repetitions, so had to manually repeat.
+const tableNameRegex = `^[\p{L}\p{M}\p{N}\p{Pc}\p{Pd}\p{Zs}]+$`
+
+var tableNameMatcher = regexp.MustCompile(tableNameRegex)
+
+// Returns nil if provided string is a valid GCP project id, else returns an
+// error. Based on rules defined in
+// [https://cloud.google.com/resource-manager/docs/creating-managing-projects].
+// Identifiers cannot be replaced via parameters in BigQuery, so interpolation
+// is a necessry evil.
+// In most cases, setting the default GCP project in the client's Query object
+// is preferable as it avoids the injection risk entirely.
+// Does not check for restricted strings such as google, null, etc.
+func ValidateGCPProjectID(projectID string) error {
+	if !projectIDMatcher.MatchString(projectID) {
+		return fmt.Errorf("invalid GCP project id")
+	}
+	return nil
+}
+
+// Returns nil if provided string is a valid dataset id, else returns an
+// error. Based on rules defined in
+// [https://cloud.google.com/bigquery/docs/datasets#dataset-naming].
+// Identifiers cannot be replaced via parameters in BigQuery, so interpolation
+// is a necessry evil.
+// In most cases, setting the default dataset in the client's Query object
+// is preferable as it avoids the injection risk entirely.
+// Does not check for restricted strings such as google, null, etc.
+func ValidateDatasetID(datasetID string) error {
+	if !datasetIDMatcher.MatchString(datasetID) {
+		return fmt.Errorf("invalid dataset id")
+	}
+	return nil
+}
```

Comment:
These feel like strange helper functions - why not just make the regex a public variable?

Reply:
I think I had two reasons:

1. The defensive java developer in me doesn't like exposing public variables, as it allows for external mutation. I get that's not really the go ethos though.

2. I wanted to document the rules I was enforcing, though I think I could just add to public var and it would be exposed by GODOC.

Reply:
Do you intend to use them in other packages? The private variable is still mutable within the package.

Reply:
I wanted to provide the option, in case someone wanted to reuse the logic, to make Public accessors to the individual validations (dataset, project, table ect)

Reply:
It would probably make more sense to have that in pkg then. If we want something re-usable, let's build it that way. But I don't think we should make something public because someone _might_ use it: https://www.hyrumslaw.com/